### PR TITLE
[HIP] [FlyDSL] dcp topk merge

### PR DESCRIPTION
**Pull request by ganyi1996ppo**
Three related changes for the Decode Context Parallel path of the DeepSeek-V3.2-style
sparse indexer. The first two are what GLM-5.2 DCP decode needs; the third is an
independent crash fix found while bringing that path up, and is useful on its own.

---
## 1. `[FlyDSL] Add DCP decode TopK merge kernel`

Under DCP each rank holds `1/W` of the KV and takes a local top-k, so the ranks must
agree on a global top-k before attention. The existing sequence reconstructs that
global set on **every** rank and then drops the `~(W-1)/W` of it the rank does not
own — W times the work, most of it discarded.

This op emits only the rank's **owned** physical KV slots, already localized and
compacted, plus the `kv_indptr` attention needs. Two kernels: a per-row radix
select for the global threshold, then a per-row pack that also computes the
cross-row prefix sum.

Two properties make the fusion possible:

- **Ownership is positional.** Candidate column `c` came from rank `c // k_loc`, so
  global ids never have to travel and the caller can exchange scores alone.
- **The slot formula collapses.** For local index `j` the physical slot is
  `block_table[j // page] * page + (j % page)` — independent of `W` and of the
  interleave size `S`. Verified by exhaustive enumeration over 50400 combinations
  of `(W, S, page, j)`, 0 mismatches. So the localize folds into the emit loop
  instead of needing a separate pass.

Allocates no device scratch (the caller owns `staging`), so it is safe inside a
captured CUDAGraph.

### Why the prefix sum lives inside `pack`

It started as its own `grid=(1,)` kernel walking the rows on a single thread. That
is O(rows) serial work sandwiched between two parallel kernels, and it showed —
**37% of the op's runtime at rows=128** (15.7 of 42.3 µs) to prefix-sum 128
integers.

Each `pack` block now recomputes the prefix it needs, in parallel. The redundancy
is real (rows blocks each scanning rows counts) but a block scan over a few hundred
values is a handful of barriers: `pack` went 3.3 → 3.5 µs while a whole kernel
launch disappeared. The loop is chunked, so `rows > 256` stays correct — verified at
rows ∈ {1, 128, 256, **257**, 512}.

That is also why `select` and `pack` cannot merge further: `pack` needs
`kv_indptr[row]`, which depends on every *other* row's count. Closing that gap needs
either a cross-block barrier (unsafe — block scheduling order is not guaranteed, so
a block spinning on a not-yet-scheduled block deadlocks) or `grid=(1,)`, which would
serialise `select` by a factor of rows.

### Performance

Against the equivalent six-op sequence, at the production shape `W=8`,
`k_loc = topk = 2048`:

| bs | six-op | fused | speedup |
|---:|-------:|------:|--------:|
| 1 | 166.5 µs | 33.9 µs | 4.91× |
| 4 | 173.6 µs | 33.8 µs | 5.14× |
| 16 | 165.2 µs | 34.3 µs | 4.82× |
| 32 | 167.3 µs | 33.6 µs | 4.98× |
| 64 | 174.6 µs | 34.3 µs | 5.09× |
| 128 | 167.4 µs | 34.0 µs | 4.92× |

The curve is **flat** — 33.6-34.3 µs regardless of bs — where the six-op path grows
with rows. Plus half the all-gather payload (`[rows, k_loc]` instead of
`[2, rows, k_loc]`) and 2 kernels instead of ~16.

This is a **single-op** measurement. It has not been attributed against a full decode
step, so the wall-clock share depends on how often the indexer runs per forward.

### Correctness

`op_tests/flydsl_tests/test_flydsl_dcp_topk_merge.py` — **45 passed**. Covers a
differential test against the six-op reference, edge cases (`topk > n_cand`, empty
rows, all-padding rows), a shape sweep, and tie-heavy inputs where the equal-to-
threshold admission rule is what decides the partition.

Two GPU races were found and fixed during development, each with a reproducible
before/after:

- **Bucket-hit test.** Every thread reads `remaining_k` from LDS but only the thread
  owning the k-th element writes it. Without a barrier between the two, a fast writer
  updates it while a slow thread is still reading, the slow thread then satisfies the
  hit condition against the *new* value and writes a second bucket — a corrupted
  threshold.
- **Block scan.** A fast wave entering the *next* `block_exclusive_prefix_i32` call
  would overwrite `scan[warp]` while a slow wave was still reading the current call's
  totals. Fixed with a trailing barrier.

The second one I initially dismissed as a non-race based on 40 clean iterations of a
single shape. It reproduced later under a different shape. A race is structural, not
a property of one observed schedule — recording that here because the first ruling
was wrong.

---

## 2. `[HIP] Add optional score output to top_k_per_row_decode`

`top_k_per_row_decode` now takes an optional `values` tensor and writes each selected
index's logit alongside it, mirroring `top_k_per_row_prefill`. The one-block kernel
already supported this (`WRITE_TOPK_VALUES` is a template parameter of
`dispatch_topk_oneblock`); the decode entry point hardcoded it to false with
`/*out=*/nullptr`. Wiring it through the header, the pybind arg list and the Python
wrapper is the whole change — **the default path compiles to the same instantiation
as before**.

This lets a DCP decode indexer drop the separate gather whose only job was to read
these same scores back out of the logits.

### Also: a padding-value fix

Where the index is padded with `-1`, the score was padded with `0.0`. A `0.0` pad does
**not** sort below real scores — logits are routinely negative, so the padding
outranks them, and a consumer that ranks these scores (a DCP merge across ranks) lets
padding steal real candidates' slots. Pad with `-inf` instead.

Measured impact before the fix: at `W=8` the disjoint owned-partition collapsed from
256 entries to 30.

The changed line is shared with prefill, but **every in-tree prefill caller passes
`values=None`** (checked all 5), so `WRITE_TOPK_VALUES` is false there and the store
never executes.

---

## Validation

- `op_tests/flydsl_tests/test_flydsl_dcp_topk_merge.py`: **45 passed**
- HIP `values`: padding `== -inf`, score `==` gathered logit, `values=None`
  bit-identical to before, non-stable path — all pass
- Gluon fix: 3 previously-aborting shapes compile; reverting the patch reproduces the
  abort on `main`, so the fix is not a no-op
- End-to-end: GLM-5.2, TP=8, DCP=8, 65536-token inputs, **8/8 requests succeeded**,
  32 prefill batches including the `done=[16384]` chunk that previously aborted. Zero
  `iota_range` / `proc died` / `exitcode=-6` in the log.

## Not covered

- **The DCP merge has not been validated on real multi-GPU DCP.** The all-gather
  itself is untested; the consumer keeps it behind a default-off flag.
- Tested on one gfx950 machine with Triton 3.7.0.
- The `waves_per_eu` timings are 10-iteration wall-clock, not a rigorous benchmark.

### CODE REVIEW

File: aiter/ops/topk.py [hunk 1]
```diff
@@ -482,6 +489,44 @@ def top_k_per_row_decode(
         k,
         workspace,
         stable,
+        values,
```
Comment (lirui927) [inline aiter/ops/topk.py:506]: should add value checks as well — dtype, shape, and contiguous.

File: op_tests/test_flydsl_dcp_topk_merge.py [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+    parser.add_argument("-p", "--page", type=int, nargs="*", default=[64])
+    parser.add_argument(
+        "-m", "--mode", type=str, nargs="*", default=CASE_MODES, choices=CASE_MODES
+    )
+    args = parser.parse_args()
+
+    df = []
+    for rows, world, k_loc, topk, page, mode in itertools.product(
+        args.rows, args.world, args.k_loc, args.topk, args.page, args.mode
+    ):
```
Comment (lirui927) [inline op_tests/test_flydsl_dcp_topk_merge.py:460]: The test organization doesn’t follow the existing CI conventions.

File: aiter/ops/flydsl/dcp_topk_merge.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+@functools.cache
+def _get_launcher(n_cand, k_loc, topk_tokens, page_size, world_size):
+    return build_dcp_topk_merge_module(
+        n_cand, k_loc, topk_tokens, page_size, world_size
+    )
+
+
+@functools.lru_cache(maxsize=8)
+def _warp_size(device_index: int) -> int:
```
Comment (coderfeli) [inline aiter/ops/flydsl/dcp_topk_merge.py]: there is get_warp_size in kernel common.py

**Comment by zufayu** [comment]
Nit (non-blocking): the commented-out `mb/ob dispatch` block at the tail of `top_k_per_row_decode` is dead code that predates this PR. Since the function is already being rewritten here, worth dropping it in a follow-up — git history keeps the reference, and leaving it invites rot (uncompiled, unmaintained, but scanned by every future reader to confirm it's unreachable).

```diff
     } else {
         aiter::ob::dispatch_topk_oneblock<float, int, 1024, false, aiter::ob::Phase::Decode>(
             ws_ptr, buf_size, logits_ptr, static_cast<int*>(nullptr),
             batch, stride0,
             /*rowStarts=*/nullptr, /*rowEnds=*/seq_lens_ptr,
             kTopK, /*out=*/nullptr, indices_ptr,
             select_min, stream, /*sorted=*/true, static_cast<int>(next_n));
     }
-
-    // --- Original mb/ob dispatch (commented out for reference) ---
-    // Both branches take the caller-provided `workspace` (no device allocation
-    // here); the mb buffer is zeroed + self_reset by the kernel (prezeroed=true).
-    // if (aiter::should_use_mulblocks(batch, stride0)) {
-    //     aiter::mb::standalone_stable_radix_topk<float, int, false, true, aiter::mb::Phase::Decode>(
-    //         ws_ptr, buf_size, logits_ptr, batch, stride0,
-    //         /*rowStarts=*/nullptr, /*rowEnds=*/seq_lens_ptr,
-    //         kTopK, /*out=*/nullptr, indices_ptr,
-    //         is_largest, stream, static_cast<int>(next_n), /*prezeroed=*/true);
-    // } else {
-    //     aiter::ob::dispatch_topk_oneblock<float, int, 1024, false, aiter::ob::Phase::Decode>(
-    //         ws_ptr, buf_size, logits_ptr, static_cast<int*>(nullptr),
-    //         batch, stride0,
-    //         /*rowStarts=*/nullptr, /*rowEnds=*/seq_lens_ptr,
-    //         kTopK, /*out=*/nullptr, indices_ptr,
-    //         select_min, stream, /*sorted=*/true, static_cast<int>(next_n));
-    // }
 }
```

Unrelated to the `values` correctness — just pre-existing hygiene that this PR's churn makes cheap to clean up.

**Review by zufayu — DISMISSED** [review]

**Review by coderfeli — DISMISSED** [review]

**Review by valarLip — DISMISSED** [review]

File: op_tests/test_flydsl_dcp_topk_merge.py [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+
+
+# @benchmark does not hide the fn from pytest: collected by its test_ name, it
+# would be called with zero fixtures and raise TypeError. The parameterized
+# wrapper below is this file's pytest entry point into the same code.
+test_dcp_topk_merge.__test__ = False
+
+
+@pytest.mark.parametrize("mode", CASE_MODES)
+@pytest.mark.parametrize("rows,world,page", [(1, 8, 64), (17, 4, 16), (256, 2, 128)])
```
Comment (valarLip) [inline op_tests/test_flydsl_dcp_topk_merge.py]: we don't use pytest.. need clean

File: op_tests/test_flydsl_dcp_topk_merge.py [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+def test_modes(rows, world, page, mode):
+    """Drive the benchmark fn's assertions over edge shapes under pytest."""
+    k_loc = 64
+    # "overtake" is the topk >= n_cand degenerate case; the rest stay selective.
+    topk = world * k_loc if mode == "overtake" else 128
+    test_dcp_topk_merge(rows, world, k_loc, topk, page, mode)
+
+
+def test_reference_model_selects_global_topk():
```
Comment (valarLip) [inline op_tests/test_flydsl_dcp_topk_merge.py]: test for reference?

**Review by valarLip — APPROVED** [review]
